crates/wasmtime/src/ref.rs

```diff
@@ -2,224 +2,179 @@
-    pub fn new(data: Box<dyn Any>) -> Self {
-        let info = AnyAndHostInfo {
-            any: data,
-            host_info: None,
-        };
-        ExternRef::Other(OtherRef(Rc::new(RefCell::new(info))))
+    /// Creates a new instance of `ExternRef` wrapping the given value.
+    pub fn new<T>(store: &crate::Store, value: T) -> ExternRef
+    where
+        T: 'static + Any,
+    {
+        let inner = VMExternRef::new(value);
+        let store = store.weak();
+        ExternRef { inner, store }
     }
 
     /// Creates a `Null` reference.
-    pub fn null() -> Self {
-        ExternRef::Null
+    pub fn null() -> Option<Self> {
+        None
     }
 
-    /// Returns the data stored in the reference if available.
-    /// # Panics
-    /// Panics if the variant isn't `ExternRef::Other`.
-    pub fn data(&self) -> cell::Ref<Box<dyn Any>> {
-        match self {
-            ExternRef::Other(OtherRef(r)) => cell::Ref::map(r.borrow(), |r| &r.any),
-            _ => panic!("expected ExternRef::Other"),
-        }
+    /// Get this reference's store.
+    ///
+    /// Returns `None` if this reference outlived its store.
+    pub fn store(&self) -> Option<crate::runtime::Store> {
+        crate::runtime::Store::upgrade(&self.store)
     }
 
-    /// Returns true if the two `ExternRef<T>`'s point to the same value (not just
-    /// values that compare as equal).
+    /// Get the underlying data for this `ExternRef`.
+    pub fn data(&self) -> &dyn Any {
+        &*self.inner
+    }
+
+    /// just point to two different values that compare as equal).
     pub fn ptr_eq(&self, other: &ExternRef) -> bool {
-        match (self, other) {
-            (ExternRef::Null, ExternRef::Null) => true,
-            (ExternRef::Ref(InternalRef(ref a)), ExternRef::Ref(InternalRef(ref b))) => {
-                a.ptr_eq(b.as_ref())
-            }
-            (ExternRef::Other(OtherRef(ref a)), ExternRef::Other(OtherRef(ref b))) => {
-                Rc::ptr_eq(a, b)
-            }
-            _ => false,
-        }
+        self.inner == other.inner
     }
 
-    /// Returns a mutable reference to the host information if available.
-    /// # Panics
-    /// Panics if `ExternRef` is already borrowed or `ExternRef` is `Null`.
-    pub fn host_info(&self) -> Option<cell::RefMut<Box<dyn Any>>> {
-        match self {
-            ExternRef::Null => panic!("null"),
-            ExternRef::Ref(r) => r.0.host_info(),
-            ExternRef::Other(r) => {
-                let info = cell::RefMut::map(r.0.borrow_mut(), |b| &mut b.host_info);
-                if info.is_none() {
-                    return None;
-                }
-                Some(cell::RefMut::map(info, |info| info.as_mut().unwrap()))
-            }
-        }
+    /// Returns the host information for this `externref`, if previously created
+    /// with `set_host_info`.
+    pub fn host_info(&self) -> Option<Rc<RefCell<dyn Any>>> {
```

Comment:
Ideally we'd move this out of the `wasmtime` crate entirely. Can all the `host_info` stuff move to `wasm_store_t` in the C API? I think the C API could store a `Rc<wasm_store_t>` on all the exported types, right?

Reply:
As mentioned in chat, it is useful to have this built into the `wasmtime` crate so that host infos can be cleaned up when their associated `externref` goes away.